solutions/gold/cf-585D.mdx

```diff
@@ -0,0 +1,144 @@
+---
+id: cf-585D
+source: CF
+title: Lizard Era Beginning
+author: Justin Ji
+---
+
+[Official Analysis](https://codeforces.com/blog/entry/20898)
+
+<Spoiler title = "Hint 1">
+
+Choosing two companions per task is equivalent to excluding one companion
+per task.
+
+</Spoiler>
+
+<Spoiler title = "Hint 2">
+
+Let's say we have a pair $(a, b)$ and $(c, d)$. If we want $a + c$ to equal $b + d$,
+then $a - b$ must be equal to $d - c$. How can this help us perform meet-in-the-middle on the
+array?
+
+</Spoiler>
+
+<Spoiler title = "Explanation">
+
+## Explanation
+
+With the information given in the hints, we can do meet-in-the-middle and consider every possible
+way to choose companions to exclude in our left and right subarrays.
+
+Given the information in hint 2, we can efficiently join results from the
+left and right sides. In a valid answer, the adjacent differences in attitudes from the
+left side must be the opposite of the adjacent differences from the right side.
+With this observation, we can find the best pairing for each possible
+result from the left side in $\mathcal{O}(\log{N})$.
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(3^{\left \lceil{\frac{N}{2}}\right \rceil}
+\log\left({3^{\left \lceil{\frac{N}{2}}\right \rceil}}\right))$
```

Comment:
shouldn't it be 
$$\mathcal{O}(3^{\left \lceil{\frac{N}{2}}\right \rceil}\log({3^{\left \lfloor{\frac{N}{2}}\right \rfloor}}))$$

also nitpick, but do you need left/right? makes the parentheses look wonky

Reply:
i requested those lol imo it makes the terms more seperable in the mind

Reply:
> shouldn't it be O(3⌈N2⌉log⁡(3⌊N2⌋))
> 
> also nitpick, but do you need left/right? makes the parentheses look wonky

hmm, I think ur right that they aren't the same, but on second thought I think it's

$\mathcal{O}(3^{\left \lfloor{\frac{N}{2}}\right \rfloor}\log\left({3^{\left \lceil{\frac{N}{2}}\right \rceil}}\right))$

because $[0, n / 2)$ has a length of floor(n / 2), and [n / 2, n] has length of ceil(n / 2).

Reply:
yeah i ran a few test cases and this is the case. will update soon